doc/subsystems/networking/net-app-api.rst

```diff
@@ -0,0 +1,207 @@
+.. _net_app_api:
+
+Network Application API
+#######################
+
+The Network Application (net-app) API allows applications to:
+
+* **Initialize** the application for networking use. This means, for example,
+  that if the application needs to have an IPv4 address, and if DHCPv4 is
+  enabled, then the net-app API will make sure that the device will get an
+  IPv4 address before the application is started.
+
+* **Set** various options for the networking subsystem. This means that if the
+  user has set options like IP addresses, IEEE 802.15.4 channel etc. in the
+  project configuration file, then those settings are applied to the system
+  before the application starts.
+
+* **Create** a simple TCP/UDP server or client application. The net-app API
+  has functions that make it easy to create a simple TCP or UDP based network
+  application. The net-app API also provides transparent TLS and DTLS support
+  for the application.
+
+The net-app API functionality is enabled by :option:`CONFIG_NET_APP` option.
+The current net-app API implementation is still experimental and may change and
+improve in future releases.
+
+Initialization
+##############
+
+The net-app API provides a cpp:func:`net_app_init()` function that can
+configure the networking subsystem for the application. Following configuration
+options control this configuration:
+
+* :option:`CONFIG_NET_APP_AUTO_INIT` automatically configures the system
+  according to other configuration options. User does not need to call
+  cpp:func:`net_app_init()` in this case as that function will be automatically
+  called when the system boots. This option is enabled by default.
+
+* :option:`CONFIG_NET_APP_INIT_TIMEOUT` specifies how long to wait for the
+  network configuration during the system boot. For example if DHCPv4 is
+  enabled, and if the IPv4 address discovery takes too long or the DHCPv4
+  server is not found, the system will resume booting after this number of
+  seconds.
+
+* :option:`CONFIG_NET_APP_NEED_IPV6` specifies that the application needs IPv6
+  connectivity. The cpp:func:`net_app_init()` function will wait until it is
+  able to setup IPv6 address for the system before continuing. This means that
+  the IPv6 duplicate address detection (DAD) has finished and the system has
+  properly setup the IPv6 address.
+
+* :option:`CONFIG_NET_APP_NEED_IPV6_ROUTER` specifies that the application
+  needs IPv6 router connectivity i.e., it needs access to external networks
+  (such as the Internet). The cpp:func:`net_app_init()` will wait until it
```

Comment:
Typing verbatim so it doesn't get interpreted as Markdown:

```
The cpp:func:`net_app_init()` function
```

(Below is also missing "function")